紀錄/25 20210420.md

```diff
@@ -0,0 +1,141 @@
+      const rightTargetBoundary = searchRightTargetBoundary(middle, right);
+      
+      const result = [leftTargerBoundary, rightTargetBoundary];
+      return result;
+    }
+    else if (nums[middle] > target) right = middle - 1;
+    else left = middle + 1;
+  }
+  return [-1, -1];
+  
+  // helper functions
+  function getMiddleIndex(left: number, right: number) {
+    if (left > right) return -1;
+    if (left === right) return left;
+    return left + Math.floor((right - left) / 2);
+  }
+  /** Search left boundary of target by Recursion */
+  function searchLeftTargetBoundary(left: number, right: number) {
+    if (nums[left] === target) return left;
+    const middle = getMiddleIndex(left, right);
+    
+    if (nums[middle] === target && nums[middle - 1] !== target) return middle;
+    else if (nums[middle] !== target) return searchLeftTargetBoundary(middle + 1, right);
+    else return searchLeftTargetBoundary(left, middle - 1);
+  }
+  /** Search right boundary of target by Recursion */
+  function searchRightTargetBoundary(left: number, right: number) {
+    if (nums[right] === target) return right;
+    const middle = getMiddleIndex(left, right);
+    
+    if (nums[middle] === target && nums[middle + 1] !== target) return middle;
+    else if (nums[middle] !== target) return searchRightTargetBoundary(left, middle - 1);
+    else return searchRightTargetBoundary(middle + 1, right);
+  }
+};
+```
+
+* MEDIUM	 162	 Find Peak Element	 https://leetcode.com/problems/find-peak-element
+
+* Template Analysis
+
+* Article Binary Search Template Analysis
+
+* 🔓	 EASY	 270	 Closest Binary Search Tree Value	 https://leetcode.com/problems/closest-binary-search-tree-value
+
+* 🔓	 MEDIUM	 702	 Search in a Sorted Array of Unknown Size	 https://leetcode.com/problems/search-in-a-sorted-array-of-unknown-size
+
+* Conclusion
+
+* MEDIUM	 50	 Pow(x, n)	 https://leetcode.com/problems/powx-n
+
+```typescript
+/** Recursion in TS by Louis. It's inspired by [the discussion](https://leetcode.com/explore/learn/card/binary-search/137/conclusion/982/discuss/19546/Short-and-easy-to-understand-solution) */
+function myPow(x: number, n: number): number {
+  if (n == 0) return 1;
+  if (n < 0) return myPow(1 / x, -n);
+  if (n % 2 == 0) return myPow(x * x, n / 2);
+  else return x * myPow(x * x, Math.floor(n / 2));
+};
+
+```typescript
+/** Iteration in TS by Louis */
+function myPow(x: number, n: number): number {
+  let base = x;
+  let exponent = n;
+  let constant = 1;
+  
+  if (n === 0) return 1;
+  
+  if (n < 0) {
+    base = 1 / x;
+    exponent = -n;
+  }
+  
+  while (exponent > 1) {
+    if (exponent % 2 == 0) {
+      base = base * base;
+      exponent = exponent / 2;
+    }
+    else {
+      constant = constant * base;
+      base = base * base;
+      exponent = Math.floor(exponent / 2);
+    }
+  }
+  
+  return constant * base;
+}
+
+
+* EASY	 367	 Valid Perfect Square	 https://leetcode.com/problems/valid-perfect-square
+
```

Comment:
```suggestion

```typescript
/** Iteration for `Valid Perfect Square` in TS by Louis*/
function isPerfectSquare(num: number): boolean {
  if (num === 1) return true;
  
  let lower = 2;
  let upper = Math.floor(num / 2);
  
  while (lower <= upper) {
    const middle = lower + Math.floor((upper - lower) / 2);
    if (num === middle * middle) return true;
    else if (num > middle * middle) lower = middle + 1;
    else upper = middle - 1;
  }
  return false;
};

```